private/bufpkg/bufimage/util.go

```diff
@@ -32,15 +32,81 @@ import (
 		return nil, err
 	}
-	// these are the files that fileOrDirPaths actually reference and will
-	// result in the non-imports in our resulting Image
-	// the Image will also include the ImageFiles that the nonImportImageFiles import
-	var nonImportImageFiles []ImageFile
+	if err := normalpath.ValidatePathsNormalizedValidatedUnique(excludeFileOrDirPaths); err != nil {
+		return nil, err
+	}
+	// These are the files that fileOrDirPaths actually reference and will
+	// result in the non-imports in our resulting Image. The Image will also include
+	// the ImageFiles that the nonImportImageFiles import
 	nonImportPaths := make(map[string]struct{})
+	var nonImportImageFiles []ImageFile
+	// We need to collect all the non-import paths in the image for a few reasons:
+	//   1. In the case where `len(fileOrDirPaths) == 0`, since all non-import paths
+	//      are now valid paths for the image, merged against the excluded paths.
+	//   2. To check against the provided paths for `allowNotExist == true` -- if this is set,
+	//      we expect all provided paths to exist to some extent.
+	allNonImportPaths := make(map[string]struct{})
+	// We have only exclude paths, and therefore all other paths are target paths.
+	if len(fileOrDirPaths) == 0 && len(excludeFileOrDirPaths) > 0 {
+		excludeFileOrDirPathMap := stringutil.SliceToMap(excludeFileOrDirPaths)
+		for _, imageFile := range image.Files() {
+			if !imageFile.IsImport() {
+				if _, ok := allNonImportPaths[imageFile.Path()]; !ok {
+					allNonImportPaths[imageFile.Path()] = struct{}{}
+				}
+				if !normalpath.MapHasEqualOrContainingPath(excludeFileOrDirPathMap, imageFile.Path(), normalpath.Relative) {
+					nonImportPaths[imageFile.Path()] = struct{}{}
+					nonImportImageFiles = append(nonImportImageFiles, imageFile)
+				}
+			}
+		}
+		// Finally, before we construct the image, we need to validate that all exclude paths
+		// provided adhere to the allowNotExist flag.
+		if !allowNotExist {
+			var pathsWithNoMatchingFiles []string
+			// TODO: right now, in order to check that each exclude exists, we need to loop through
+			// all image files for each excludeFileOrDirPaths to return an error that provides
+			// a list of paths that are not found within the image. This seems like a pretty expensive
+			// operation, but seems important for surfacing a coherent error to the user.
+			for _, excludeFileOrDirPath := range excludeFileOrDirPaths {
+				var foundPath bool
+				for nonImportPath := range allNonImportPaths {
+					if normalpath.EqualsOrContainsPath(excludeFileOrDirPath, nonImportPath, normalpath.Relative) {
+						foundPath = true
+						break
+					}
+				}
+				if !foundPath {
+					pathsWithNoMatchingFiles = append(pathsWithNoMatchingFiles, excludeFileOrDirPath)
+				}
+			}
+			if len(pathsWithNoMatchingFiles) > 0 {
+				return nil, fmt.Errorf("paths with no matching files in the image: %v", pathsWithNoMatchingFiles)
+			}
```

Comment:
In the `--path` handling below, it looks like we only report the first offending file (not all of them). For simplicity and symmetry, feel free to do the same here (and below). For reference, this is what we have for `--path`:

```
	if !allowNotExist {
		for potentialDirPath := range potentialDirPathMap {
			if _, ok := matchingPotentialDirPathMap[potentialDirPath]; !ok {
				// no match, this is an error given that allowNotExist is false
				return nil, fmt.Errorf("path %q has no matching file in the image", potentialDirPath)
			}
		}
```